torchrl/envs/transforms/transforms.py

```diff
@@ -6796,3 +6795,211 @@ def _reset(
             tensordict_reset = self._call(tensordict_reset)
         return tensordict_reset
+
+
+class HERTransform(Transform):
+    """Hindsight Experience Replay (HER) transform.
+
+    This transform is used in reinforcement learning algorithms that employ
+    Hindsight Experience Replay (HER). HER is a technique that allows an agent
+    to learn from failed experiences by replaying them with different goals.
+
+    Args:
+        samples (Optional[Union[int, torch.Tensor]]): The number of augmented samples
+            to generate for each original sample. Defaults to 4.
+        generation_type (str): The type of goal generation to use. Can be one of
+            "future", "random", or "final". Defaults to "future".
+        achieved_goal_key (Optional[NestedKey]): The key to access the achieved goal
+            in the input tensor dictionary. Defaults to "achieved_goal".
+        desired_goal_key (Optional[NestedKey]): The key to access the desired goal
+            in the input tensor dictionary. Defaults to "desired_goal".
+        reward_key (Optional[NestedKey]): The key to access the reward in the output
+            tensor dictionary. Defaults to "reward".
+        reward_function (Optional[callable]): The reward function to use for calculating
+            the rewards of augmented samples. Defaults to None, in which case the
+            `distance_reward_function` is used.
+
+    Attributes:
+        ENV_ERR (str): The error message to raise when the transform is applied to
+            the collector or the environment.
+
+    """
+
+    ENV_ERR = (
+        "The Reward2GoTransform is only an inverse transform and can "
+        "only be applied to the replay buffer and not to the collector or the environment."
+    )
+
+    def __init__(
+        self,
+        samples: Optional[Union[int, torch.Tensor]] = 4,
+        generation_type: str = "future",
+        achieved_goal_key: Optional[NestedKey] = "achieved_goal",
+        desired_goal_key: Optional[NestedKey] = "desired_goal",
+        reward_key: Optional[NestedKey] = "reward",
+        reward_function: Optional[callable] = None,
+    ):
+        super().__init__(
+            in_keys=None,
+            in_keys_inv=None,
+            out_keys_inv=None,
+        )
+        self.achieved_goal_key = achieved_goal_key
+        self.desired_goal_key = desired_goal_key
+        self.reward_key = reward_key
+        self.generation_type = generation_type
+
+        if reward_function is None:
+            self.reward_function = distance_reward_function
+        else:
+            self.reward_function = reward_function
+
+        if not isinstance(samples, torch.Tensor):
+            samples = torch.tensor(samples)
+
+        self.register_buffer("samples", samples)
+
+    def _inv_call(self, tensordict: TensorDictBase) -> TensorDictBase:
+        augmentation_td = self.her_augmentation(tensordict)
+        return torch.cat([tensordict, augmentation_td], dim=0)
```

Comment:
Why not a transform? I think it's pretty neat to use a transform. Who said a transform had to change things in-place?
Our API to modify samples at writing time is to use either a transform or a different writer. If you think this can be achieved with a writer I'm on board. But I don't think there's anything wrong with the transform.

An advantage of using a writer instead os that it feels more natural (transforms can be used with envs unless specified otherwise, writers are dedicated to RBs)